rand_core/src/lib.rs

```diff
@@ -212,6 +188,81 @@ pub trait RngCore {
 /// [`BlockRngCore`]: block::BlockRngCore
 pub trait CryptoRng: RngCore {}
 
+/// The potentially fallible core of a random number generator.
+pub trait TryRngCore {
+    /// The type returned in the event of a RNG error.
+    type Error: fmt::Debug + fmt::Display;
+
+    /// Return the next random `u32`.
+    fn try_next_u32(&mut self) -> Result<u32, Self::Error>;
+    /// Return the next random `u64`.
+    fn try_next_u64(&mut self) -> Result<u64, Self::Error>;
+    /// Fill `dest` entirely with random data.
+    fn try_fill_bytes(&mut self, dst: &mut [u8]) -> Result<(), Self::Error>;
+
+    /// Wrap RNG with the [`PanicOnErr`] wrapper.
+    fn panic_on_err(self) -> UnwrapErr<Self>
+    where Self: Sized {
+        UnwrapErr(self)
+    }
+
+    /// Convert an [`RngCore`] to a [`RngReadAdapter`].
+    #[cfg(feature = "std")]
+    fn read_adapter(&mut self) -> RngReadAdapter<'_, Self>
+    where Self: Sized {
+        RngReadAdapter { inner: self }
+    }
+}
+
+/// A marker trait used to indicate that an [`TryRngCore`] implementation is
+/// supposed to be cryptographically secure.
+///
+/// See [`CryptoRng`] docs for more information.
+pub trait TryCryptoRng: TryRngCore {}
+
+impl<R: RngCore> TryRngCore for R {
+    type Error = Infallible;
+
+    #[inline]
+    fn try_next_u32(&mut self) -> Result<u32, Self::Error> {
+        Ok(self.next_u32())
+    }
+
+    #[inline]
+    fn try_next_u64(&mut self) -> Result<u64, Self::Error> {
+        Ok(self.next_u64())
+    }
+
+    #[inline]
+    fn try_fill_bytes(&mut self, dst: &mut [u8]) -> Result<(), Self::Error> {
+        self.fill_bytes(dst);
+        Ok(())
+    }
+}
+
+impl<R: CryptoRng> TryCryptoRng for R {}
```

Comment:
These blanket impls are good, except for the conflicts with other impls we want.

Should we instead add a macro like: `impl_try_rng_from_rng_core!(ChaCha8Rng);` ?

If specialization ever gets implemented, we could add this impl back while replacing the macro with a deprecated placeholder to avoid causing much actual breakage.

Reply:
This prevents us only from implementing `Box<R>` and `&mut R`, which I think is less important for IO-based RNGs. For example, in RustCrypto we decided to use `&mut impl CryptoRng` because in our opinion it's less confusing for users.

Ideally, we should not need those blanket impls in the first place.

Reply:
On a second thought, the macro could be a fine (albeit inelegant) stopgap solution.